### ISSUE TITLE
[ros2param] [ros2lifecycle] Crash with OpenSplice

### ISSUE THREAD

**Issue by mikaelarguedas** [ros2/ros2cli#107]
When testing https://github.com/ros2/ros2cli/pull/106 I noticed I cannot get `ros2 param` or `ros2 lifecycle` on opensplice. `ros2 service` works though:

### How to reproduce
Shell 1:
`ros2 run demo_nodes_cpp parameter_node`
Shell 2:
`ros2 param set parameter_node foo 42`

### Actual behavior 
```
Failed to wait on wait set: failed to detach guard condition, at /home/mikael/work/ros2/bouncy_ws/src/ros2/rmw_opensplice/rmw_opensplice_cpp/src/rmw_wait.cpp:375, at /home/mikael/work/ros2/bouncy_ws/src/ros2/rcl/rcl/src/rcl/wait.c:647
```

### Expected behavior
```
Set parameter successful
```

Get the same error message when doing service calls via `ros2 lifecycle`
```
ros2 run lifecycle lifecycle_talker
```
```
ros2 lifecycle get lc_talker
```

**Comment by sloretz** [comment]
Same issue on windows 10

```
>ros2 param list talker
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Fri Jun 22 13:47:04 Pacific Daylight Time 2018
Node        : DESKTOP-S5KSNBT
Process     : python.exe <1492>
Thread      : 1292
Internals   : WaitSet.cpp/456/6.7.180404OSS/5926d257/5926d257/-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529700424.164618200
Failed to wait on wait set: failed to detach guard condition, at C:\J\workspace\packaging_windows\ws\src\ros2\rmw_opensplice\rmw_opensplice_cpp\src\rmw_wait.cpp:375, at C:\J\workspace\packaging_windows\ws\src\ros2\rcl\rcl\src\rcl\wait.c:647
```

**Comment by dirk-thomas** [comment]
How about the following workaround: ros2/rmw_opensplice#235?

**Comment by wjwwood** [comment]
What precondition is not met? I feel that if we're going to accept that workaround, we either need to extend the check to know which precondition it is (and the decide if that's ok to ignore) or we need to know there's only one precondition and that it's ok to ignore.

**Comment by dirk-thomas** [comment]
What precondition is not met?

The not met precondition is mentioned in the error message:

> This ReadCondition is being deleted

**Comment by wjwwood** [comment]
Sorry, missed that in the message, but the second half of my comment still stands, I think. Can we detect which precondition is not met? If not, is this the only precondition for this version of OpenSplice that cannot be met? If no to both, then I think it's pretty dangerous to ignore all unmet preconditions.

**Comment by dirk-thomas** [comment]
In the guard condition this seems to be the only case: https://github.com/ADLINK-IST/opensplice/blob/36e4f9c250a640ade72d5af92084f5f5243ce96c/src/api/dcps/c%2B%2B/common/code/GuardCondition.cpp#L101-L104

And no, we can't check the same condition from the outside: https://github.com/ADLINK-IST/opensplice/blob/36e4f9c250a640ade72d5af92084f5f5243ce96c/src/api/dcps/c%2B%2B/common/include/Condition.h#L40-L42

**Comment by wjwwood** [comment]
Then maybe this check is ok, though it sucks that if they return precondition unmet for something else in the future we'd miss that. But that's the best we can do?

However, a more generic question is why do we get in this state? Maybe it's done from the signal handler of OpenSplice?

Also we could check `deinitializing` if we inherited from the `Condition` (it's only ever set `true` by subclasses AFAICT).

I'm still a bit wary that we're treating a symptom and not the cause.

**Comment by dhood** [comment]
On xenial https://github.com/ros2/rmw_opensplice/pull/235 made it so that the list and get commands do something now, but the error is still printed. In the case of setting a lifecycle node's state, it is printing 22 errors (reliably).

<details><summary>Full commands/error output</summary>

<p>

```
$ ros2 run lifecycle lifecycle_talker 
```
```
$ ros2 lifecycle set lc_talker configure
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:32 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13283>
Thread      : main thread 7f3759327700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960012.714741194
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:32 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13283>
Thread      : main thread 7f3759327700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960012.715479737
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:32 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13283>
Thread      : main thread 7f3759327700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960012.721998394
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:32 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13283>
Thread      : main thread 7f3759327700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960012.722054236
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:32 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13283>
Thread      : main thread 7f3759327700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960012.732224140
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:32 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13283>
Thread      : main thread 7f3759327700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960012.732262321
Transitioning successful


$ ros2 lifecycle set lc_talker activate
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:42 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13336>
Thread      : main thread 7f0e267e7700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960022.201973860
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:42 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13336>
Thread      : main thread 7f0e267e7700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960022.202410846
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:42 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13336>
Thread      : main thread 7f0e267e7700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960022.208401851
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:42 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13336>
Thread      : main thread 7f0e267e7700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960022.208473021
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:42 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13336>
Thread      : main thread 7f0e267e7700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960022.682554190
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:42 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13336>
Thread      : main thread 7f0e267e7700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960022.682612383
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:42 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13336>
Thread      : main thread 7f0e267e7700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960022.683281934
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:42 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13336>
Thread      : main thread 7f0e267e7700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960022.683341593
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:43 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13336>
Thread      : main thread 7f0e267e7700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960023.183078300
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:43 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13336>
Thread      : main thread 7f0e267e7700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960023.183127718
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:43 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13336>
Thread      : main thread 7f0e267e7700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960023.183703261
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:43 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13336>
Thread      : main thread 7f0e267e7700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960023.183739744
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:43 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13336>
Thread      : main thread 7f0e267e7700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960023.683527461
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:43 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13336>
Thread      : main thread 7f0e267e7700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960023.683568903
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:43 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13336>
Thread      : main thread 7f0e267e7700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960023.684161231
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:43 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13336>
Thread      : main thread 7f0e267e7700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960023.684199360
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:44 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13336>
Thread      : main thread 7f0e267e7700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960024.183981611
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:44 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13336>
Thread      : main thread 7f0e267e7700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960024.184028841
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:44 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13336>
Thread      : main thread 7f0e267e7700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960024.184642176
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:44 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13336>
Thread      : main thread 7f0e267e7700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960024.184687525
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:44 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13336>
Thread      : main thread 7f0e267e7700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960024.208775744
========================================================================================
Context     : DDS::WaitSet::detach_condition
Date        : Mon Jun 25 13:53:44 PDT 2018
Node        : osrf-esteve
Process     : python3.5 <13336>
Thread      : main thread 7f0e267e7700
Internals   : WaitSet.cpp/456/6.7.180404OSS///-1
----------------------------------------------------------------------------------------
Report      : Precondition not met: This ReadCondition is being deleted.
Internals   : DDS::OpenSplice::ReadCondition::wlReq_detachFromWaitset/ReadCondition.cpp/176/4/1529960024.208852766
Transitioning successful
```

</p>
</details>

---

Since it's printing so many errors I would be reluctant to consider this issue resolved.

Two observations:
- The `on_activate` callback [takes 2 seconds](https://github.com/ros2/demos/blob/bfd63e19dbcecd77febd8f1dea573a6d3e047d94/lifecycle/src/lifecycle_talker.cpp#L157)
- Calling the underlying services with `ros2 service` does not cause the errors to be output.

```
$ ros2 service call /lc_talker/change_state lifecycle_msgs/ChangeState "{node_name: lc_talker, transition: {id: 1}}"
requester: making request: lifecycle_msgs.srv.ChangeState_Request(node_name='lc_talker', transition=lifecycle_msgs.msg.Transition(id=1, label=''))

response:
lifecycle_msgs.srv.ChangeState_Response(success=True)

$ ros2 service call /lc_talker/change_state lifecycle_msgs/ChangeState "{node_name: lc_talker, transition: {id: 3}}"
requester: making request: lifecycle_msgs.srv.ChangeState_Request(node_name='lc_talker', transition=lifecycle_msgs.msg.Transition(id=3, label=''))

response:
lifecycle_msgs.srv.ChangeState_Response(success=True)
```

From this I suspect this is caused by how `ros2 lifecycle` /`ros2 param` are calling the services.

**Comment by dirk-thomas** [comment]
> Since it's printing so many errors I would be reluctant to consider this issue resolved.

This issue - as the title says - is about a crash: `Crash with OpenSplice`. That specific problem has been "resolved" by ros2/rmw_opensplice#235. Since the workaround is not satisfactory a follow up ticket has been created: ros2/rmw_opensplice#237. Therefore I do think this ticket should be closed.

**Comment by dhood** [comment]
sure, we can move this to a separate issue. However it is that we want to capture that there's still work to be done
